Make filter property animatable.

mantaroh opened the pull request:
<!-- Please describe your changes on the following line: -->
This is a PR for https://bugzilla.mozilla.org/show_bug.cgi?id=1362897


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

<!-- Either: -->
There are tests for these changes, a test case will be landed in web-platform-tests in https://bugzilla.mozilla.org/show_bug.cgi?id=1362897

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17288)
<!-- Reviewable:end -->

mantaroh committed:
Preserve the unit when interpolating/adding angles with matching units.

If the units of two angles being interpolated/added matches, we should preserve
the original unit; otherwise, we fall back to radians. This matches the behavior
of Gecko.

mantaroh committed:
Add function which convert from URLValueData to SpecifiedUrl.

We will need to convert from URLValueData to SpecifiedUrl when cloning the
stylo's data from gecko's data. Because the filter structure on gecko has url as
URLValue type. We will use this convert function when interpolating as
discrete also(Bug 1368610).

mantaroh committed:
Make filter property animatable.

mantaroh committed:
Add compute distance for Filter.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/effects.mako.rs, components/style/gecko/url.rs, components/style/properties/gecko.mako.rs, components/style/properties/helpers/animated_properties.mako.rs
 * @emilio: components/style/properties/longhand/effects.mako.rs, components/style/gecko/url.rs, components/style/properties/gecko.mako.rs, components/style/properties/helpers/animated_properties.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

mantaroh commented:
This patche have been reviewed by @birtles, @heycam, @hiikezoe, @BorisChiou

mantaroh commented:
Ah,  I removed unnecessary line before sending PR in order to build servo.(Because servo don't need the `use properties::longhands::filter::computed_value::Filter::*;`)
But this modification made unresolved the DropShadow when building the geckolib.

nox commented:
> Ah, I removed unnecessary line before sending PR in order to build servo.(Because servo don't need the `use properties::longhands::filter::computed_value::Filter::*;`)
> But this modification made unresolved the DropShadow when building the geckolib.

You have to qualify the variants name by `Filter::` instead of glob-importing all of those enum's variants.

nox commented on components/style/properties/gecko.mako.rs:
```diff
@@ -3518,6 +3512,64 @@ fn static_assert() {
+                            (**filter.__bindgen_anon_1.mDropShadow.as_ref()).mArray[0];
+                        shadow_val = orig_shadow.to_shadow();
+                    };
+                    filters.push(DropShadow(shadow_val));
```
AFAICT you could just do:

```rust
filters.push(unsafe {
    (**filter.__bindgen_anon_1.mDropShadow.as_ref()).mArray[0].to_shadow()
});
```

mantaroh replied:
Thanks, @nox !

I addressed it, and changed to use `Filter::` in order to qualify the variant name.

nox commented:
@bors-servo r=birtles,heycam,hiikezoe,BorisChiou,nox

bors-servo commented:
:pushpin: Commit c91f67d has been approved by `birtles,heycam,hiikezoe,BorisChiou,nox`

<!-- @bors-servo r=birtles,heycam,hiikezoe,BorisChiou,nox c91f67d9aec08d6b3e85e01227b24d1c86e870a1 -->

bors-servo commented:
:hourglass: Testing commit c91f67d9aec08d6b3e85e01227b24d1c86e870a1 with merge 07f1060e9af285ca712b4e8bb1acf0013068adb5...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/293)

jdm commented:
```
  ▶ TIMEOUT [expected PASS] /css-animations-1_dev/html/css-filters-animation-blur.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-brightness.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-combined-001.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-contrast.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected PASS] /css-animations-1_dev/html/css-filters-animation-drop-shadow.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-grayscale.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-hue-rotate.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-invert.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-opacity.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-saturate.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ TIMEOUT [expected FAIL] /css-animations-1_dev/html/css-filters-animation-sepia.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1
```

emilio commented:
Presumably those tests were broken before, and animations were just not running. The story to stop the compositor to take a screenshot in Servo when animations have finished is not great :(

mantaroh commented:
I thought this cause is simply forgetting to remove meta file, but I faced test timeout even if remove meta file. I'm not sure why servo will be TIMEOUT..
If I understand correct, this reftest take a screenshot when load event firing, it looks like servo don't fire the load event..

I tried the `reftest-wait`, however the result is same timeout.

mantaroh commented:
Umm,, If remove animation, reftest are success, and it is success when animation has finished. It looks like that unfinished animation caused reftest timeout. 

@emilio , @SimonSapin , @Manishearth 
The reftest with unfinished animation will cause timeout in Servo.
Do you know something about this phenomenon?

mantaroh committed:
Add the disable anotation of filter's css-tests in order to skip timeout fail.

mantaroh commented:
I want to separate this problem into another issue. I filed the issue #17335, in this PR, I'll skip those tests.

hiikezoe commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4da244261557f8c505316bd907ff0d9b2e843e45 with merge 9520af5a61fde0380e90ea85fe927265dd102454...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4612)

mantaroh commented:
@bors-servo retry

bors-servo commented:
@mantaroh: :key: Insufficient privileges: and not in try users

birtles commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 4da244261557f8c505316bd907ff0d9b2e843e45 with merge 3a5a028889f1fda7db01e89b679583e236fe80e4...

birtles commented:
Looks like they're still timing out though.

hiikezoe commented:
yeah, other tests in tests/wpt/css-tests/filters-1_dev.

jdm commented on tests/wpt/metadata-css/css-animations-1_dev/html/css-filters-animation-brightness.htm.ini:
```diff
@@ -1,3 +1,3 @@
 [css-filters-animation-brightness.htm]
   type: reftest
-  expected: FAIL
+  disabled: The reftest of servo with non-expired animations failed(issue 17335).
```
There is no need to disable these tests. We should use `expected: TIMEOUT` instead, and add a separate `bug: https://github.com/servo/servo/issues/17335` as well.

mantaroh commented:
Thanks, @hiikezoe , @birtles , @jdm .
I overlooked that I need to change the ini file of filters-1_dev.
I updated init file of tests/wpt/css-tests/filters-1_dev and tests/wpt/css-tests/css-animations-1_dev by using `expected: TIMEOUT`.

birtles commented:
@bors-servo retry

birtles commented:
@bors-servo p=10

Manishearth commented:
@bors-servo r=birtles

bors-servo commented:
:pushpin: Commit ac75e0e has been approved by `birtles`

<!-- @bors-servo r=birtles ac75e0ec4788d31eadcb34e03f213bfa14993dd1 -->

birtles commented:
@bors-servo r=hiro

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #17353

bors-servo commented:
:pushpin: Commit ac75e0e has been approved by `hiro`

<!-- @bors-servo r=hiro ac75e0ec4788d31eadcb34e03f213bfa14993dd1 -->

bors-servo commented:
:hourglass: Testing commit ac75e0ec4788d31eadcb34e03f213bfa14993dd1 with merge bceae420479e9a15f5ea52f70a0ca0a3edd12f3e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4643)

birtles commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit ac75e0ec4788d31eadcb34e03f213bfa14993dd1 with merge 1721fa20603fecfe64ef9da1477181eb27be674c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4621)

jdm commented:
```
  ▶ FAIL [expected TIMEOUT] /css-animations-1_dev/html/css-filters-animation-contrast.htm
  └   → /css-animations-1_dev/html/css-filters-animation-contrast.htm 0f0e0e7166718073fb1271121f89e4aecfbc750d
/css-animations-1_dev/html/reference/css-filters-animation-contrast-ref.htm 7e54b8d874be52ec34436f49bd8b3230e7f70e17
Testing 0f0e0e7166718073fb1271121f89e4aecfbc750d == 7e54b8d874be52ec34436f49bd8b3230e7f70e17

  ▶ PASS [expected TIMEOUT] /filters-1_dev/html/css-filters-animation-drop-shadow.htm
```
Oh good, these animation filter tests give us intermittent results, too!

jdm commented:
I have filed issues so that they will be classified as intermittent failures. You can retry this PR without any changes.

birtles commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit ac75e0ec4788d31eadcb34e03f213bfa14993dd1 with merge 29483eb14a5d03e5de00447edd5149af38ffe0fd...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4635)

birtles commented:
Looks like a few more intermittents?

jdm commented:
Updated.

birtles commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit ac75e0ec4788d31eadcb34e03f213bfa14993dd1 with merge 244d8aaa42c5a846fabd5f5ff3c7c59ba80a053c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4651)

jdm commented:
Nevermind. It's clearly going to be a laborious process to keep the tests enabled; just disable them all instead (both the filter animation ones in /css-animations-1_dev/html/ and /filters-1_dev/html/).

mantaroh commented:
Thanks @jdm, @birtles 
I updated those tests meta file with the disabled annotation.

birtles commented:
I'm not sure if we use exactly the same format for these meta files as we do in Gecko, but if so I guess we just need a `__dir__.ini` that specifies `disabled: https://github.com/servo/servo/issues/17335`
(As per https://developer.mozilla.org/en-US/docs/Mozilla/QA/web-platform-tests#Per-directory_settings)

Also, let's drop the unrelated changes to MANIFEST.json

mantaroh commented:
Thanks @birtles 
In those tests directory contain another tests, so I updated the disable annotation's comment, and removed the changes of metadata-css/MANIFEST.json.

birtles commented:
@bors-servo delegate+

bors-servo commented:
:v: @mantaroh can now approve this pull request

mantaroh commented:
@bors-servo retry

mantaroh commented:
@bors-servo r=hiro

bors-servo commented:
:pushpin: Commit d32948c has been approved by `hiro`

<!-- @bors-servo r=hiro d32948c46abc6e37e1c341801e94873c96154db7 -->

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7553), [arm32](http://build.servo.org/builders/arm32/builds/7769), [arm64](http://build.servo.org/builders/arm64/builds/7808), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7896), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4653), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4678), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7879), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/364), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/363), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4583), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4649), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/359), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/362), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3125)
Approved by: hiro
Pushing e41b7d06b473544755902be2c1334a7622c84940 to master...
